### PR TITLE
Fix: verification of incremental correctness that was not working because of using wrote writeFile

### PR DESCRIPTION
**Pull request by sheetalkamat**
This ensures that we use sys.writeFile for baseline buildinfo so that buildInfo correctness is verified. It was missing when we started passing the writeFileCallback which didnt track written files. 
This also baselines those differences so its easy to view, read and maintain.

This is on top of #48703. Actual change is only 386786c and f45e9ca

### CODE REVIEW

File: src/testRunner/unittests/tsbuild/noEmitOnError.ts [hunk 1]
```diff
@@ -8,40 +8,81 @@ namespace ts {
...
+                noChangeRun,
+                {
+                    subScenario: "Fix error",
+                    modifyFs: fs => fs.writeFileSync("/src/src/main.ts", `import { A } from "../shared/types/db";
 const a = {
     lastName: 'sdsd'
-};`, "utf-8")
-        );
+};`, "utf-8"),
+                    descripencyExplaination: noChangeWithExportsDescripencyRun.descripencyExplaination,
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/noEmitOnError.ts:43]: Typo: "Discrepancy"
Typo: "Explanation"

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 2]
```diff
@@ -328,136 +287,173 @@ interface Symbol {
         generateBundleFileSectionInfo(sys, originalReadCall || sys.readFile, baselineRecorder, bundle.dts, declarationFilePath);
         baselineRecorder.Close();
         const text = baselineRecorder.lines.join("\r\n");
-        (originalWriteFile || sys.writeFile).call(sys, `${buildInfoPath}.baseline.txt`, text);
+        sys.writeFile(`${buildInfoPath}.baseline.txt`, text);
     }
-
-    interface VerifyTscEditCorrectnessInput {
+    interface VerifyTscEditDescripanciesInput {
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/helpers.ts:292]: Typo: "Discrepancies"

Comment (amcasey) [reply src/testRunner/unittests/tsbuild/helpers.ts:292]: And below

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 3]
```diff
@@ -328,136 +287,173 @@ interface Symbol {
...
-    }
+            else {
+                if (actual !== undefined) return;
+            }
+            writeIncorretness(actual, expected, message);
+        }
 
-    function verifyPresenceAbsence<T>(actual: T | undefined, expected: T | undefined, message: string) {
-        (expected !== undefined ? assert.isDefined : assert.isUndefined)(actual, message);
+        function writeIncorretness<T>(actual: T | undefined, expected: T | undefined, message: string) {
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/helpers.ts:440]: Typo: "Incorrectness"

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 4]
```diff
@@ -503,51 +499,34 @@ interface Symbol {
         modifyFs: (fs: vfs.FileSystem) => void;
         subScenario: string;
         commandLineArgs?: readonly string[];
-        cleanBuildDiscrepancies?: () => ESMap<string, CleanBuildDescrepancy>;
+        descripencyExplaination?: () => readonly string[];
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/helpers.ts:502]: Why an array, rather than a single string?

Comment (amcasey) [reply src/testRunner/unittests/tsbuild/helpers.ts:502]: And why wrap in a function?  Is the string expensive to compute?

File: src/testRunner/unittests/tsc/helpers.ts [hunk 5]
```diff
@@ -10,7 +10,15 @@ namespace ts {
         subScenario: "no-change-run",
         modifyFs: noop
     };
+    export const noChangeWithExportsDescripencyRun: TestTscEdit = {
+        ...noChangeRun,
+        descripencyExplaination: () => [
+            "Incremental build does not change and has .ts as signature so exports has all imported modules/referenced files",
```
Comment (amcasey) [inline src/testRunner/unittests/tsc/helpers.ts:16]: Does "Incremental build does not change" mean "Incremental build inputs did not change"?

File: src/testRunner/unittests/tsc/incremental.ts [hunk 6]
```diff
@@ -372,11 +375,10 @@ declare global {
                 {
                     subScenario: "Delete output for class3",
                     modifyFs: fs => fs.unlinkSync("/src/projects/project1/class3.d.ts"),
-                    cleanBuildDiscrepancies: () => new Map<string, CleanBuildDescrepancy>([
-                        // Ts buildinfo willbe updated but will retain lib file errors from previous build and not others because they are emitted because of change which results in clearing their semantic diagnostics cache
-                        // But in clean build because of global diagnostics, semantic diagnostics are not queried so not cached in tsbuildinfo
-                        ["/src/projects/project2/tsconfig.tsbuildinfo", CleanBuildDescrepancy.CleanFileTextDifferent]
-                    ]),
+                    descripencyExplaination: () => [
+                        "Ts buildinfo willbe updated but will retain lib file errors from previous build and not others because they are emitted because of change which results in clearing their semantic diagnostics cache",
```
Comment (amcasey) [inline src/testRunner/unittests/tsc/incremental.ts:379]: Typo: "will be"

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 7]
```diff
@@ -328,136 +287,173 @@ interface Symbol {
...
+        }
+        if (!headerAdded && descripencyExplaination) addBaseline("*** Supplied descripency explaination but didnt file any difference");
+        return baselines;
 
-            function verifyTextEqual(incrementalText: string | undefined, cleanText: string | undefined, descrepancyInClean: CleanBuildDescrepancy | undefined, message: string) {
-                if (descrepancyInClean === undefined) {
-                    assert.equal(incrementalText, cleanText, message);
-                    return;
+        function verifyTextEqual(incrementalText: string | undefined, cleanText: string | undefined, message: string) {
+            if (incrementalText !== cleanText) writeIncorretness(incrementalText, cleanText, message);
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/helpers.ts:398]: Typo: "Incorrectness"

**Review by amcasey — COMMENTED** [review]
So the tests used to call the unpatched `writeFile` and that was always wrong.  They were, apparently, doing this because the assert was too strict?

File: src/testRunner/unittests/tsbuild/helpers.ts (hunk 4 above)
Comment (sheetalkamat) [reply src/testRunner/unittests/tsbuild/helpers.ts:502]: These are inputs when the testcase is created so this will ensure calculation only if needed

Comment (sheetalkamat) [reply src/testRunner/unittests/tsbuild/helpers.ts:502]: Array just helps keep lines even as we are anyways joining the strings letter

Comment (amcasey) [reply src/testRunner/unittests/tsbuild/helpers.ts:502]: The calculation we're postponing is an array of string literals?  If you're going to keep the array, please add a comment about how to interpret the elements.

**Comment by amcasey** [comment]
So the tests used to call the unpatched writeFile and that was always wrong? They were, apparently, doing this because the assert was too strict?

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 8]
```diff
@@ -461,34 +499,21 @@ interface Symbol {
         modifyFs: (fs: vfs.FileSystem) => void;
         subScenario: string;
         commandLineArgs?: readonly string[];
-        cleanBuildDiscrepancies?: () => ESMap<string, CleanBuildDescrepancy>;
+        discrepancyExplanation?: () => readonly string[];
```
Comment (amcasey) [inline src/testRunner/unittests/tsbuild/helpers.ts:502]: Maybe a comment like "An array of lines to be printed in order when a discrepancy is detected"?

**Review by amcasey — COMMENTED** [review]
LGTM, assuming I have correctly understood why calling the original `writeFile` was a mistake.

**Pull request merged**